File: spec/linkedList.js
import chai, { expect } from 'chai'
import chaiChange from 'chai-change'
// import Node from '../src/node'
import LinkedList from '../src/linkedList'

chai.use(chaiChange)

describe('LinkedList', function() {
  'use strict'

  it('exists', function() {
    expect(LinkedList).to.be.a('function')
  })

  context('getHeadNode()', function() {
    it('returns null when there is no head node', function() {
      const myLinkedList = new LinkedList()
      expect(myLinkedList.getHeadNode()).to.equal(null)
    })

    it('returns the head node', function() {
      const myLinkedList = new LinkedList()
      myLinkedList.insertFirst('hello')

      expect(myLinkedList.getHeadNode().getData()).to.equal('hello')
    })
  })


})

      // expect(() => myLinkedList.push('foo'))
      //   .to.alter(() => myLinkedList.length(), { from: 0, to: 1 })

File: java_implementation/LinkedList.java
package datastructures;

public class LinkedList {

  private Node head;
  private int length = 0;

  private class Node {
    private final Object data;
    private Node next;

    Node(Object data) {
      this.data = data;
    }
  }

  public Object getHeadNode() {
    return this.head.data;
  }

  public Object getTailNode() {
    Node currentNode = this.head;
    while (currentNode.next != null) {
      currentNode = currentNode.next;
    }
    return currentNode.data;
  }

  public Node insert(Object data) {
    Node node = new Node(data);
    if (this.head == null) {
      this.head = node;
      this.length++;
      return node;
    }
    Node currentNode = this.head;
    while (currentNode.next != null) {
      currentNode = currentNode.next;
    }
    currentNode.next = node;
    this.length++;
    return node;
  }

  public int size() {
    return this.length;
  }
}